### ISSUE TITLE
Transient CI failures

### ISSUE THREAD

**Issue by cburstedde** [cburstedde/libsc#213]
There are rather frequent transient CI failures. Even rerunning several times resolves them only slowly. I'm wondering if there's anything in our powers to reduce or eleminate these. Are we sure that there are no race conditions involving file access?

**Comment by tim-griesbach** [comment]
> Is there a particular workflow (CI for CMake, CI for Autotools) that is failing intermittently? Is it a specific test (like 'scda') or subset of tests that's failing?

Based on my observations, it is only the `scda` test that causes transient CI failures. The CI failures seem to be due to unexpected MPI count values retrieved from MPI staus objects, i.e. a MPI I/O function completes without an error but reports a lower count of written/read bytes. This causes an scda count error (`SC_SCDA_FERR_COUNT`) and hence the CI fails. 

> Are we sure that there are no race conditions involving file access?

I didn't spot anything along these lines on the level of  MPI_File calls.

In p4est, we have a similar situation in `test_loadsave{2,3}`, e.g. https://github.com/cburstedde/p4est/actions/runs/12528069581. These tests use `p4est_{load,save}`, which perform MPI I/O directly through MPI_File functions, bypassing the `sc_io` file function wrappers for legacy reasons. Despite this difference, I observed analogous behavior: no MPI I/O errors, yet occasional incorrect MPI I/O counts during the CI runs. To sum it up, we have a substantially different MPI I/O workflow in p4est's `test_loadsave{2,3}` that leads to very similar transient CI fails as for scda.

In summary, it seems likely to me that the file system in GitHub Actions is contributing to the transient CI failures. One approach to mitigate the issue could be to implement a loop in the respective `sc_io` wrapper functions to repeat the I/O call until the sum of reported counts matches the expected count. We could introduce a maximal number of loop passes and maybe there is a reasonable number of maximal loop passes to avoid the described issue.

**Comment by cburstedde** [comment]
Thanks all for the analysis.

> > Is there a particular workflow (CI for CMake, CI for Autotools) that is failing intermittently? Is it a specific test (like 'scda') or subset of tests that's failing?
> 
> Based on my observations, it is only the `scda` test that causes transient CI failures. The CI failures seem to be due to unexpected MPI count values retrieved from MPI staus objects, i.e. a MPI I/O function completes without an error but reports a lower count of written/read bytes. This causes an scda count error (`SC_SCDA_FERR_COUNT`) and hence the CI fails. 
> 
> In p4est, we have a similar situation in `test_loadsave{2,3}`, e.g. https://github.com/cburstedde/p4est/actions/runs/12528069581. These tests use `p4est_{load,save}`, which perform MPI I/O directly through MPI_File functions, bypassing the `sc_io` file function wrappers for legacy reasons. Despite this difference, I observed analogous behavior: no MPI I/O errors, yet occasional incorrect MPI I/O counts during the CI runs. To sum it up, we have a substantially different MPI I/O workflow in p4est's `test_loadsave{2,3}` that leads to very similar transient CI fails as for scda.
> 
> In summary, it seems likely to me that the file system in GitHub Actions is contributing to the transient CI failures. One approach to mitigate the issue could be to implement a loop in the respective `sc_io` wrapper functions to repeat the I/O call until the sum of reported counts matches the expected count. We could introduce a maximal number of loop passes and maybe there is a reasonable number of maximal loop passes to avoid the described issue.

I think this may be harmless and useful.  We might just keep writing when at
least 1 byte was processed.  On a count of 0, we should return an error.  Is
this feasible in a minimally invasive way?  Please go ahead @tgriesbach.

**Comment by tim-griesbach** [comment]
> I think this may be harmless and useful. We might just keep writing when at least 1 byte was processed. On a count of 0, we should return an error. Is this feasible in a minimally invasive way? Please go ahead @tgriesbach.

I tested the approach of trying to write/read again if the retrieved count does not match the expected count. This approach does not seem to be a solution to the issue since I could only observe that the the counts are either 0 or as expected. In particular, an erroneous count of 0 persisted even after 50 retries, indicating that retrying the I/O operations does not overcome the observed CI failures.

However, I did notice that introducing sleep calls after every I/O operations resulted consistently in a successful CI. I think this supports the hypothesis that the CI failures may be due to a busy file system. I will work on a patch that can be applied during the CI to include sleep calls where it is necessary for a more reliable CI.

**Comment by tim-griesbach** [comment]
> However, I did notice that introducing sleep calls after every I/O operations resulted consistently in a successful CI. I think this supports the hypothesis that the CI failures may be due to a busy file system. I will work on a patch that can be applied during the CI to include sleep calls where it is necessary for a more reliable CI.

I implemented this approach in https://github.com/cburstedde/libsc/issues/213 and it seems to solve the issue.

**Comment by cburstedde** [comment]
Let's close this for now and periodically re-check whether the github CI file system works for us again.
